ToolChain/Assembler/Parser/Parser.py

```diff
@@ -20,586 +19,393 @@
 __license__ = "GPL"
-__version__ = "2.0"
+__version__ = "2.1"
 __maintainer__ = "CSE"
 __status__ = "Dev"
 
+from ToolChain.Assembler.Constants import REGISTERS, \
+                                          INSTRUCTION_LIST, \
+                                          LABEL_INSTRUCTIONS, \
+                                          STATE_LIST, \
+                                          MEMORY_REFERENCE_INDICATORS, \
+                                          EXPORTED_REFERENCE_INDICATOR, \
+                                          DATA_NUMERIC_INDICATOR, \
+                                          DATA_ALPHA_INDICATOR, \
+                                          DATA_MEMORY_REFERENCE, \
+                                          COMMENT_INDICATORS, \
+                                          IMMEDIATE_PREFIX, \
+                                          REGISTER_PREFIX, \
+                                          FLAGS_INDICATORS, \
+                                          WIDTH_INDICATORS
+
+from CapuaEnvironment.IntructionFetchUnit.FormDescription import formDescription
+from CapuaEnvironment.Instruction.OperationDescription import operationDescription
+
+import struct
+import re
+
 
 class Parser:
     """
     This class is used to parse the text format code and build a list of instruction from it.
     To do so, it has a direct link into CapuaEnvironment instruction information. Those class
     and description files are directly used in order to help build the binary code required
     to run code inside of the Capua environment. The steps are as follows:
-
     1. Verify if the file name is valid, and that it is not empty.
     2. Read the file and parse each line into a list to be evaluated one at a time.
     3. Evaluate each line and, if not an empty line, store instructions and memory references in lists.
-         - Parse each line and determine whether we're dealing with an instruction, memory reference, data, etc.
-         - If an instruction is found, determine all possible forms for this instruction (eg. ins - reg - reg)
-         - Using possible forms for an instruction, determine the correct instruction code based on given arguments.
+        - Parse each line and determine whether we're dealing with an instruction, memory reference, data, etc.
+        - If an instruction is found, determine all possible forms for this instruction (eg. ins - reg - reg)
+        - Using possible forms for an instruction, determine the correct instruction code based on given arguments.
     4. Once all info has been evaluated, if correct, we generate the binary code for each instruction
     """
 
-    loadAddress = 0x00
-    instructionList = None
-    fileContent = None
-    referenceDict = None
-    finalSize = 0
+    relativeAddressCounter = 0                # Used to determine memory address of a label at a given point in the file
+    lineno = 0                                # Used to pinpoint where ValueErrors are raised
```

Comment:
The parser parses a single line of code at a time. The line number should be the responsibility of the assembler. If the Parser raises exceptions, then the assembler should catch these and be able to generate the proper error message.